BacktraceScreen.c

```diff
@@ -0,0 +1,405 @@
+/*
+htop - BacktraceScreen.h
+(C) 2023 htop dev team
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "config.h" // IWYU pragma: keep
+
+#include "BacktraceScreen.h"
+
+#if defined(HAVE_BACKTRACE)
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+
+#include <ncursesw/curses.h>
+
+#include "CRT.h"
+#include "FunctionBar.h"
+#include "ListItem.h"
+#include "Macros.h"
+#include "Object.h"
+#include "Panel.h"
+#include "Platform.h"
+#include "Process.h"
+#include "RichString.h"
+#include "Settings.h"
+#include "Vector.h"
+#include "XUtils.h"
+
+#define BACKTRACE_FRAME_HEADER_INDEX -1
+
+#define BACKTRACE_PANEL_OPTIONS_NAME_DEMANGLE "Demangle"
+#define BACKTRACE_PANEL_OPTIONS_NAME_RAW "Raw"
+
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL "Full Path"
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_BASENAME "Basename"
+
+#define BACTRACE_PANEL_HEADER_NUMBER_FRAME "#"
+#define BACTRACE_PANEL_HEADER_ADDRESS "ADDRESS"
+#define BACTRACE_PANEL_HEADER_NAME "NAME"
+#define BACTRACE_PANEL_HEADER_PATH "PATH"
+
+#define MAX_ADDR_32 (0xffffffff)
+#define MAX_HEX_ADDR_STR_LEN_32 (strlen("0x") + 8)
+#define MAX_HEX_ADDR_STR_LEN_64 (strlen("0x") + 16)
+
+static const char* const BacktraceScreenFunctions[] = {
+   "Refresh",
+#if defined(HAVE_DEMANGLING)
+   BACKTRACE_PANEL_OPTIONS_NAME_RAW,
+#endif
+   BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL,
+   "Done   ",
+   NULL
+};
+
+static const char* const BacktraceScreenKeys[] = {
+   "F1",
+#if defined(HAVE_DEMANGLING)
+   "F2",
+#endif
+   "F3",
+   "Esc",
+   NULL
+};
+
+static const int BacktraceScreenEvents[] = {
+   KEY_F(1),
+#if defined(HAVE_DEMANGLING)
+   KEY_F(2),
+#endif
+   KEY_F(3),
+   27,
+};
+
+static void BacktraceFrame_displayHeader(const BacktraceFrame* const frame, RichString* out) {
+   const BacktracePanelPrintingHelper* const printingHelper = &frame->backtracePanel->printingHelper;
+   assert(printingHelper);
+
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   int maxFunctionNameLength = printingHelper->maxFunctionNameLength;
+   if ((displayOptions & DEMANGLE_NAME_FUNCTION) == DEMANGLE_NAME_FUNCTION &&
+         printingHelper->maxDemangledFunctionNameLength > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFunctionNameLength;
+   }
+
+   char* line = NULL;
+   int len = xAsprintf(&line, "%-*s %-*s %-*s %-*s",
+                       (int)printingHelper->maxNumberFrameLength, BACTRACE_PANEL_HEADER_NUMBER_FRAME,
+                       (int)printingHelper->maxAddressLength, BACTRACE_PANEL_HEADER_ADDRESS,
+                       (int)maxFunctionNameLength, BACTRACE_PANEL_HEADER_NAME,
+                       (int)printingHelper->maxObjectPathLength, BACTRACE_PANEL_HEADER_PATH);
+
+   RichString_appendnAscii(out, CRT_colors[DEFAULT_COLOR] | A_ITALIC | A_BOLD | A_DIM, line, len);
```

Comment:
Flags should be handled via named color constants in `CRT.c`. Using `CRT_colors[] | flag …` should be the exception.

Reply:
It's the second header, and I don't know how to manage it correctly. Currently, it is just a frame with the index `-1`, and it works. For the colors, it's the only method I found. I tried `CRT_colors[FUNCTION_BAR]` however the color didn't take up all the line space, even if I used the `w` of `Panel`.

> Using CRT_colors[] | flag … should be the exception.

I know it's used by `LinuxProcess`, so can it stay here?

Reply:
I added a color called `BACTRACE_HEADER`.